File: ppmdemo/rust/src/board.rs
use std::env;
use std::fmt::Write as FmtWrite;
use std::fs;
use std::io::BufWriter;
use std::io::Write as IoWrite;

pub fn generate() {
    let scale = [16, 9];
    let fps = 60;
    let cell = fps;
    let w = scale[0] * cell;
    let h = scale[1] * cell;
    let rgb_max = 0xFF;
    let root = env::var("CARGO_MANIFEST_DIR").unwrap();
    let output_dir = root + "/output/board";
    if !fs::exists(&output_dir).unwrap() {
        fs::create_dir_all(&output_dir).unwrap();
    }
    let mut buf = String::new();
    let header = format!("P6\n{} {}\n{}\n", w, h, rgb_max);
    for i in 0..fps {
        FmtWrite::write_fmt(&mut buf, format_args!("{}/frame-{:02}.ppm", output_dir, i)).unwrap();
        let output_path = &buf;
        let file = fs::File::create(output_path).unwrap();
        let mut file = BufWriter::new(file);
        IoWrite::write_all(&mut file, header.as_bytes()).unwrap();
        for y in 0..h {
            for x in 0..w {
                let mut buf = [0x00, 0x00, 0x00];
                if ((x + i) / cell + (y + i) / cell) % 2 == 1 {
                    buf[0] = 0xFF;
                }
                IoWrite::write_all(&mut file, &buf).unwrap();
            }
        }
        println!("Generated {}", output_path);
        buf.clear();
    }
}

File: ppmdemo/c/board.c
#include <stdio.h>

int main() {
    char buf[256];
    int fps = 60;
    int w = 16 * fps;
    int h = 9 * fps;
    for (int i = 0; i < fps; i++) {
        snprintf(buf, sizeof(buf), "output_%02d.ppm", i);
        const char* output_path = buf;
        FILE* file = fopen(output_path, "wb");
        fprintf(file, "P6\n");           // format
        fprintf(file, "%d %d\n", w, h);  // width / height
        fprintf(file, "255\n");          // the maximum value of the color component, usually 255 for P6
        for (int y = 0; y < h; y++) {
            for (int x = 0; x < w; x++) {
                if (((x + i) / fps + (y + i) / fps) % 2) {
                    fputc(0xFF, file);
                } else {
                    fputc(0x00, file);
                }
                fputc(0x00, file);
                fputc(0x00, file);
            }
        }
        fclose(file);
        printf("Generated %s\n", output_path);
    }
    return 0;
}
